### ISSUE TITLE
Bad shape check in matrix multiplication raises ValueError

### ISSUE THREAD

**Issue by benoit-bnp** [cvxpy/cvxpy#2065]
**Describe the bug**
When running the resolution of a CVXPY problem, a ValueError is raised by CVXPY when performing matrix multiplication due to how shapes of input factors are checked.

**To Reproduce**
Please download [this file](https://github.com/cvxpy/cvxpy/files/10988554/cvxpy_error_vars.npz.zip) containing NumPy matrices (`.npz` file inside the `.zip`).
```python
import numpy as np
import cvxpy as cvx

k = 0.23
t = 0
ptf = 0.25
npzfile = np.load('cvxpy_error_vars.npz')

var_w = cvx.Variable(npzfile['sigma'].shape[0])
var_r = npzfile['mu_t'] @ var_w
r2 = cvx.quad_form(var_w, npzfile['sigma'])
var_te = cvx.quad_form(var_w - npzfile['benchmark'], npzfile['sigma'])
var_i = cvx.pnorm(cvx.sqrt(npzfile['omega']) @ var_w, 2)

objective = var_r - k * var_i - t * var_te
objective -= npzfile['lambda_value'] @ r2

sd_constraint = r2 <= ptf ** 2
constraints = [
    var_w <= npzfile['max_expo'],
    npzfile['min_expo'] <= var_w,
    cvx.sum(var_w) == 1,
    sd_constraint,
]

prob = cvx.Problem(cvx.Maximize(objective), constraints)
print(prob.solve(solver='ECOS', max_iters=100, abstol=1e-10, reltol=1e-10, feastol=1e-10))
```

**Expected behavior**
The resolution of the problem: should print `0.009980184128397372`.

**Output**
```
Traceback (most recent call last):
  File "<project_path>/min_cvxpy_error.py", line 27, in <module>
    print(prob.solve(solver='ECOS', max_iters=100, abstol=1e-10, reltol=1e-10, feastol=1e-10))
  File "<project_path>/venv/lib/python3.10/site-packages/cvxpy/problems/problem.py", line 493, in solve
    return solve_func(self, *args, **kwargs)
  File "<project_path>/venv/lib/python3.10/site-packages/cvxpy/problems/problem.py", line 1068, in _solve
    self.unpack_results(solution, solving_chain, inverse_data)
  File "<project_path>/venv/lib/python3.10/site-packages/cvxpy/problems/problem.py", line 1398, in unpack_results
    self.unpack(solution)
  File "<project_path>/venv/lib/python3.10/site-packages/cvxpy/problems/problem.py", line 1347, in unpack
    self._value = self.objective.value
  File "<project_path>/venv/lib/python3.10/site-packages/cvxpy/problems/objective.py", line 97, in value
    v = self.args[0].value
  File "<project_path>/venv/lib/python3.10/site-packages/cvxpy/atoms/atom.py", line 357, in value
    return self._value_impl()
  File "<project_path>/venv/lib/python3.10/site-packages/cvxpy/atoms/atom.py", line 371, in _value_impl
    arg_val = arg._value_impl()
  File "<project_path>/venv/lib/python3.10/site-packages/cvxpy/atoms/atom.py", line 371, in _value_impl
    arg_val = arg._value_impl()
  File "<project_path>/venv/lib/python3.10/site-packages/cvxpy/atoms/atom.py", line 376, in _value_impl
    result = self.numeric(arg_values)
  File "<project_path>/venv/lib/python3.10/site-packages/cvxpy/atoms/affine/binary_operators.py", line 115, in numeric
    return np.matmul(values[0], values[1])
ValueError: matmul: Input operand 1 does not have enough dimensions (has 0, gufunc core with signature (n?,k),(k,m?)->(n?,m?) requires 1)
```

**Version**
 - OS: macOS 13.2.1
 - CVXPY Version: 1.3.0

**Additional context**
I'm updating some old code from CVXPY 0.4.11 to latest 1.3.0 so I might miss something ? Otherwise, what fixed it for me was to change [this line that checks shapes](https://github.com/cvxpy/cvxpy/blob/v1.3.0/cvxpy/atoms/affine/binary_operators.py#L111) to

```python
if values[0].shape == () or values[1].shape == () or \
           intf.is_sparse(values[0]) or intf.is_sparse(values[1]):
```

**Comment by SteveDiamond** [comment]
@benoit-bnp I have incorporated your fix into #2067. The bigger issue is that quad_form here has dimensions (1,1), which makes no sense. But we can't change that behavior in a patch IMO, so will address in 1.4.

**Comment by benoit-bnp** [comment]
Thanks for the quick fix !
